### PR TITLE
Add timeout for dial

### PR DESCRIPTION
**Pull request by howardjohn**
This avoids hanging forever when we fail to connect

### CODE REVIEW

**Review by nmittler — APPROVED** [review]

File: pkg/istio-agent/xds_proxy.go [hunk 1]
```diff
@@ -171,15 +171,17 @@ func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDisc
 	defer close(stop)
 	go p.healthChecker.PerformApplicationHealthCheck(healthEventsChan, stop)
 
-	upstreamConn, err := grpc.Dial(p.istiodAddress, p.istiodDialOptions...)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	upstreamConn, err := grpc.DialContext(ctx, p.istiodAddress, p.istiodDialOptions...)
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/xds_proxy.go:176]: Will this influence the backoff reconnect?

Comment (hzxuzhonghu) [reply pkg/istio-agent/xds_proxy.go:176]: //Once this function returns, the cancellation and expiration of
// ctx will be noop. 

got the answer

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**